### DISCUSSION TITLE
@GeneratedColumn

### DISCUSSION THREAD

**Discussion by gavinking** [hibernate/hibernate-orm#4526] [Design Proposals]
We have a `@ColumnDefault` annotation which works nicely in conjunction with `@Generated(INSERT)`. But some databases have a syntax like `generated as (....)` which is a bit more powerful than `default`.

In https://github.com/hibernate/hibernate-orm/pull/4220/ I tried supporting this by adding an `as` member to `@Generated`, but perhaps that's not the right place to put it. So what are the other options? A new annotation, `@ColumnGeneratedAs` or `@GeneratedAs`?

Suggestions?

**Comment by gavinking** [comment]
Another closely related issue: I'm starting to think it's silly to have to write `@Generated(INSERT) @ColumnDefault` and `@Generated(ALWAYS) @ColumnGeneratedAlways`.

I think we could make:

- `@ColumnDefault` imply `@Generated(INSERT)` and
- `@ColumnGeneratedAlways` imply `@Generated(ALWAYS)`.

I know it's a bit out of character to have the DDL annotations imply stuff at the object level rather than the other way around, but it makes perfect sense in this case.

**Comment by gavinking** [comment]
After discussion with @sebersole we've identified two remaining issues:

### The name

`@ColumnGeneratedAlways` is bad. 

I now lean toward `@GeneratedColumn`, but alternatives include `@GeneratedAlways` or `@AlwaysGenerated`.

### Multiple-column mappings

This annotation is currently&mdash;like its friend `@ColumnDefault`&mdash;limited to single-column mappings.

There are a range of imperfect solutions to this:

1. Disambiguate the column by name

       @GeneratedColumn(name="total", value="quantity*unitPrice")

   I feel like this is a bit fragile if you have stuff like naming strategies, but it's in-principle more powerful since you can use it in conjunction with `@JoinColumn`, `@DiscriminatorColumn`, etc.
2. Embed a `@Column` annotation

       @GeneratedColumn(column=@Column(name="total"), value="quantity*unitPrice")

   This is less fragile but also less flexible.
3. Copy/paste all the members of `@Column` into `@GeneratedColumn`. Icky, but convenient for the user.

Note that what we decide to do here for `@GeneratedColumn` also needs to work for `@ColumnDefault`. On balance I think option 1 is best, as long as the `name` was optional for single-column mappings.

**Comment by gavinking** [comment]
After going down a somewhat deeper rabbit hold than anticipated, and discovering that there are reasonable ways to use `@ColumnDefault` (some documented, some not, and none covered by the tests) that really don't fit into the same usage pattern as `@GeneratedColumn`, I've decided not to shoehorn together these two annotations.

**Comment by sebersole** [comment]
At first glance, I think it only makes sense to allow this on basic attributes.  On that premise, I think the ability to name the column is unnecessary.  Also on that premise, I think the annotation actually should be geared toward the attribute rather than the column.  As you say, it describes aspects of both an typically we "push these down" to the db level.  Given all of that, I'd lean towards:

```
@AlwaysGenerated( "quantity*unitPrice" )
@Column( name="total" )
```

**Comment by sebersole** [comment]
Though to match better with the phrasing on the SQL side,  `@GeneratedAlways` might be better

**Comment by gavinking** [comment]
@sebersole I totally missed your comments here.

Do you want me to *remove* `name` from these annotations (I can totally do it, it just means deleting code).
